server/app/middlewares/websocket_backend.rb

```diff
@@ -236,19 +267,30 @@ def our_version
 
   # @param [Faye::WebSocket] ws
   # @param [HostNode] node
-  def handle_invalid_agent_version(ws, node)
-    node.set(connected: false, last_seen_at: Time.now.utc)
-    EventMachine::Timer.new(1) do
-      ws.close(4010) if ws
-    end
+  def handle_invalid_agent_version(ws, node, version)
+    send_master_info(ws)
+    ws.close(4010, "agent version #{version} is not compatible with server version #{Server::VERSION}")
+  end
+
+  # Send master_info RPC notification directly, without looping through the normal RPC mechanisms
+  #
+  # @param [Faye::Websocket] ws
+  def send_master_info(ws)
+    # symbols in RPC parameters are implicitly converted into strings by MongoPubsub
+    send_rpc_notify(ws, '/agent/master_info', {'version' => Server::VERSION})
   end
 
   # @param [Faye::Websocket] ws
+  def send_rpc_notify(ws, method, *params)
+    send_message(ws, [2, method, params])
+  end
+
+  # Must be called in EM thread.
+  #
+  # @param [Faye::Websocket] ws
   # @param [Array] message
   def send_message(ws, message)
-    EM.next_tick {
-      ws.send(MessagePack.dump(message).bytes)
-    }
+    ws.send(MessagePack.dump(message).bytes)
```

Comment:
You can always do some monkey patching during specs if private methods hinder you speccing:
```
subject.singleton_class.class_eval{public :foo}
expect(subject.foo).to eq('foo')
```

Of course that's not too nice and leaves the possibility of specs passing but actual code failing to call private methods.